Sources/EpoxyCollectionView/CollectionView/ReusableViews/CollectionViewCell.swift

```diff
@@ -120,22 +120,31 @@ public final class CollectionViewCell: UICollectionViewCell, ItemCellView {
     // On iOS 15, cells that return an unstable size can result in a layout recursion crash. The
     // root cause of these crashes is likely an ambiguous layout due to misconfigured constraints.
     // Unfortunately, finding each and every one of these ambiguous layouts can be challenging. This
-    // code works around this new behavior / crash by giving up on self-sizing after 5 attempts.
-    // Once a component has been sized differently 5 times, we assume that it has an ambiguous
-    // layout and we simply return the last computed size for the remainder of the component's life.
-    // We reset the count on cell reuse and when the cell is reconfigured in `CollectionView`'s
-    // `configure` function.
-    if CollectionViewConfiguration.shared.enableLayoutRecursionWorkaround {
-      if let previousComputedSize = previousComputedSize, numberOfNewComputedSizes >= 5 {
-        EpoxyLogger.shared.assertionFailure(
-          "Layout recursion detected. View: \(view?.description ?? "nil view"). Size: \(preferredAttributes.size).")
-        preferredAttributes.size = previousComputedSize
+    // code tries to detect these layout issues and logs a warning when one is detected: if a
+    // component size is unstable for at least 50 sizing attempts in less than 2 seconds, we log a
+    // warning.
+
+    let sizingAttemptTime = CACurrentMediaTime()
+    let wasFirstSizingAttemptRecent: Bool
+    if let firstSizingAttemptTime = firstSizingAttemptTime {
+      wasFirstSizingAttemptRecent = sizingAttemptTime - firstSizingAttemptTime < 2
+      if !wasFirstSizingAttemptRecent {
+        // If the last sizing attempt was not nil, but also more than the 2s threshold, reset all
+        // self-sizing-layout-recursion-tracking state.
+        resetSelfSizingLayoutRecursionTrackingState()
       }
+    } else {
+      firstSizingAttemptTime = sizingAttemptTime
+      wasFirstSizingAttemptRecent = false
+    }
 
-      if preferredAttributes.size != previousComputedSize {
-        numberOfNewComputedSizes += 1
-        previousComputedSize = preferredAttributes.size
-      }
+    if wasFirstSizingAttemptRecent, previousComputedSizes.count >= 50 {
+      EpoxyLogger.shared.warn(
+        "Layout recursion detected. View: \(view?.description ?? "nil view"). Sizes: \(previousComputedSizes).")
```

Comment:
I don't think this will get measured into our logging systems the same way assertions are, can we continue to use an assertion? Or are you thinking about re-mapping this warning into an assertion in our logging adapter?

Reply:
Ah ok, I'll change this back to being an assert